src/planner.cc

```diff
@@ -663,175 +668,205 @@ void Planner::Impl::bring_path_to_halt() {
+    return false;
+  }
+
+  const uint32_t steps = planned.TotalSteps();
+  const uint32_t steps_to_vmax =
+    uniformly_accelerated_ramp_distance(start_speed, planned.v1, target.accel);
+
+  // We intersect with the travel part.
+  if (steps_to_vmax <= planned.travel) {
+    planned.v0 = start_speed;
+    planned.accel = steps_to_vmax;
+    planned.travel = planned.travel - steps_to_vmax;
+    return true;
+  }
+
+  // We don't intersect with travel.
+  planned.travel = 0;
+  const double end_speed =
+    uniformly_accelerated_ramp_speed(steps, start_speed, target.accel);
+
+  // We do only accel and decel, no travel.
+  if (end_speed > planned.v2) {
+    // NOTE(lromor): Here we are searching for the amount of steps at which
+    // two accel-decel profiles would converge. If this value ends up being
+    // very close to the end length of the segment, it means that we might
+    // have a single step dedicated to deceleration. Due to step
+    // discretization we have a rounding error that mighe undershoot or
+    // overshoot the actual joining speed for instance going below the final
+    // speed v2 or above v1. We fix this by casting to int and assuming that
+    // by doing so, the worst case scenario is the new v1 to go below v2. At
+    // that point we force v2 to be equal v1 and remove any other deceleration
+    // step.
+    planned.accel = find_uniformly_accelerated_ramps_intersection(
+      start_speed, planned.v2, target.accel, steps);
+    planned.v1 = uniformly_accelerated_ramp_speed(planned.accel, start_speed,
+                                                  target.accel);
+    if (planned.v1 <= planned.v2) {
+      planned.v2 = planned.v1;
+      planned.accel = steps;
+    }
+
+    planned.decel = steps - planned.accel;
+    planned.v0 = start_speed;
+    return true;
+  }
+
+  // At the end of the segment we fall below v2, means
+  // we remove any travel and decel.
+  planned = {steps, 0, 0, start_speed, end_speed, end_speed};
+  return true;
+}
+
 // Perform a backward and forward pass across
 // the whole planning buffer to adapt the start
 // and final speed of all the segments.
 void Planner::Impl::UpdateMotionProfile() {
   assert(!planning_buffer_.empty());
 
   // Starting speed for each Planning segment.
-  PlanningSegment *segment;
   int buffer_index = planning_buffer_.size() - 1;
 
   // Next speed is the initial speed of the following(next) segment on the
   // original defining axis. We start from the backward pass so it's always 0
   // since we plan to always decelerate to zero.
-  double speed = 0;
-  GCodeParserAxis defining_axis = planning_buffer_.back()->target.defining_axis;
+  PlanningSegment *next_segment = nullptr;
+  PlanningSegment *previous_segment = nullptr;
 
-  // Perform the backward pass. Compute and join the travel and deceleration
-  // parts of the profile.
+  // Perform a backward pass.
   for (; buffer_index >= last_planned_invariant_segment_; --buffer_index) {
```

Comment:
For a follow-up PR it might be worthwhile to explore if we can just iterate through the buffer (I think currently the container does not have an iterator yet, so that would need to be added). For later.